### PR TITLE
lmtest::coeftest

### PR DESCRIPTION
**Pull request by vincentarelbundock**
this is paired with a commit to the `insight` repo.

``` r
library(lmtest)
library(parameters)
library(insight)

mod1 <- lm(hp ~ mpg, mtcars)
mod2 <- coeftest(mod1)

standard_error(mod2)
#>     Parameter        SE
#> 1 (Intercept) 27.433296
#> 2         mpg  1.309585
p_value(mod2)
#>     Parameter            p
#> 1 (Intercept) 8.245456e-13
#> 2         mpg 1.787835e-07
ci(mod2)
#>     Parameter    CI_low    CI_high
#> 1 (Intercept) 268.05605 380.108579
#> 2         mpg -11.50426  -6.155202
degrees_of_freedom(mod2)
#> [1] 30
get_statistic(mod2)
#>     Parameter Statistic
#> 1 (Intercept) 11.813466
#> 2         mpg -6.742389
model_parameters(mod2, ci=.9)
#> Warning: Could not get model data.
#> Parameter   | Coefficient |    SE |           90% CI | t(30) |      p
#> ---------------------------------------------------------------------
#> (Intercept) |      324.08 | 27.43 | [277.52, 370.64] | 11.81 | < .001
#> mpg         |       -8.83 |  1.31 | [-11.05,  -6.61] | -6.74 | < .001
```

### CODE REVIEW

File: R/model_parameters.other.R [hunk 1]
```diff
@@ -121,3 +121,22 @@ model_parameters.glmx <- function(model,
...
+
+
+# lmtest::coeftest -------------------------------------------------------
+
+#' @export
+model_parameters.coeftest <- function(model,
+                                      ci = .95,
+                                      verbose = TRUE,
+                                      ...) {
+  out <- parameters:::.model_parameters_generic(
```
Comment (strengejacke) [inline R/model_parameters.other.R:133]: You should remove `parameters:::` here.

File: R/ci.R [hunk 2]
```diff
@@ -994,6 +994,19 @@ ci.meta_bma <- ci.meta_random
 
 
 
+# lmtest::coeftest -------------------------------------------------------
+
+#' @export
+ci.coeftest <- function(model, ci=0.95, vcov.=NULL, df=NULL, ...) {
+  out <- confint(model, level=ci, vcov.=vcov., df=df)
```
Comment (strengejacke) [inline R/ci.R:1001]: Please use `stats::confint()` here.

File: R/ci.R [hunk 3]
```diff
@@ -994,6 +994,19 @@ ci.meta_bma <- ci.meta_random
 
 
 
+# lmtest::coeftest -------------------------------------------------------
+
+#' @export
+ci.coeftest <- function(model, ci=0.95, vcov.=NULL, df=NULL, ...) {
```
Comment (strengejacke) [inline R/ci.R:1000]: I think you can remove the arguments `vcov` and `df` here, as these are only used inside `coefci()`, while `confint()` is applied to the coeftest-object that already has processed the `vcov`-argument.

**Review by strengejacke — CHANGES_REQUESTED** [review]

File: R/ci.R [hunk 4]
```diff
@@ -994,6 +994,19 @@ ci.meta_bma <- ci.meta_random
...
+
+#' @export
+ci.coeftest <- function(model, ci=0.95, ...) {
+  out <- stats::confint(model, level=ci)
+  out <- .data_frame(
+    Parameter = .remove_backticks_from_string(row.names(out)),
+    CI_low = out[, 1],
+    CI_high = out[, 2]
+  )
+  row.names(out) <- NULL
```
Comment (strengejacke) [inline R/ci.R:1007]: You actually don't need to set rownames to NULL here, because this is done in `.data_frame()`.

**Review by strengejacke — APPROVED** [review]

**Comment by strengejacke** [comment]
Thanks, looks good!

**Comment by strengejacke** [comment]
@vincentarelbundock Thanks for the PR. I have made some minor revisions to the `ci()` method (https://github.com/easystats/parameters/commit/7df9a146a4dbefe9b84f4a508522d28bd1d028bf), which I will explain here, to help you understand the API / design / internals.

If possible, it is useful to add a `degrees_of_freedom()` method for an object, if there is a special handling. Else, the default method may return the correct df, or `Inf`. If it fails, it might be useful to add a `degrees_of_freedom()` method that just returns `Inf`. In this case, I added following code:

```r
#' @export
degrees_of_freedom.coeftest <- function(model, ...) {
  attributes(model)$df
}
```

When the CIs for an object are simply based on _estimate +/- 1.96 * SE_, then you can call the `ci_wald()` method. That's why `ci.coeftest` is now simply:

```r
#' @export
ci.coeftest <- ci.lm
```

`ci.lm()` calls `ci_wald()` with the degrees of freedom set to `NULL`, which means the df are retrieved via `degrees_of_freedom()`. And this is what actually `confint()` returns for `coeftest`:

``` r
library(lmtest)
#> Loading required package: zoo
#> 
#> Attaching package: 'zoo'
#> The following objects are masked from 'package:base':
#> 
#>     as.Date, as.Date.numeric
library(parameters)
mod1 <- lm(hp ~ mpg, mtcars)
m <- coeftest(mod1)

# CI assuming normal distribution...
cbind(
  coef(m) - standard_error(m)$SE * qnorm(.975),
  coef(m) + standard_error(m)$SE * qnorm(.975)
)
#>                  [,1]       [,2]
#> (Intercept) 270.31404 377.850586
#> mpg         -11.39647  -6.262992

# same as t-distribution with Inf degrees of freedom
cbind(
  coef(m) - standard_error(m)$SE * qt(.975, df = Inf),
  coef(m) + standard_error(m)$SE * qt(.975, df = Inf)
)
#>                  [,1]       [,2]
#> (Intercept) 270.31404 377.850586
#> mpg         -11.39647  -6.262992

# same as ci_wald with dof = Inf
ci_wald(m, dof = Inf)
#>     Parameter CI    CI_low    CI_high
#> 1 (Intercept) 95 270.31404 377.850586
#> 2         mpg 95 -11.39647  -6.262992

# object has 30 df
degrees_of_freedom(m)
#> [1] 30

# confint takes df into account
confint(m, level = .95)
#>                 2.5 %     97.5 %
#> (Intercept) 268.05605 380.108579
#> mpg         -11.50426  -6.155202

# we can compute by hand with t-distribution and 30 df
cbind(
  coef(m) - standard_error(m)$SE * qt(.975, df = 30),
  coef(m) + standard_error(m)$SE * qt(.975, df = 30)
)
#>                  [,1]       [,2]
#> (Intercept) 268.05605 380.108579
#> mpg         -11.50426  -6.155202

# or call ci_wald with specific df resp. auto-detect df
ci_wald(m, dof = NULL) # auto-detect df via degrees_of_freedom
#>     Parameter CI    CI_low    CI_high
#> 1 (Intercept) 95 268.05605 380.108579
#> 2         mpg 95 -11.50426  -6.155202
```

<sup>Created on 2020-12-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Long story short: if a model-object uses `get_parameters(m)$Estimate - standard_error(m)$SE * qt(.975, df = df)` to compute standard errors, you can simply call `ci_wald(model)` - but you have to make sure that `degrees_of_freedom()` is available, if necessary.

**Comment by vincentarelbundock** [comment]
Got it. Very useful, thanks!
